tests/unit_tests/steps/test_compile_fortran.py

```diff
@@ -109,129 +109,164 @@ def test_vanilla(self, compiler):
-
-        assert SOURCE_CHANGED in recompile_reasons
-        assert FLAGS_CHANGED in recompile_reasons
-
-
-# todo: test compile_file here? it's just glue
-
-# todo: test write_compile_result & read_compile_result here - perhaps as integration tests?
+    def ensure_mods_stored(self, mock_copy, mods_combo_hash='eac3b22d'):
+        # make sure the newly created mod files were copied TO the prebuilds folder
+        mock_copy.assert_has_calls(
+            calls=[
+                call(Path('/fab/proj/build_output/mod_def_1.mod'),
+                     Path(f'/fab/proj/build_output/_prebuild/mod_def_1.{mods_combo_hash}.mod')),
+                call(Path('/fab/proj/build_output/mod_def_2.mod'),
+                     Path(f'/fab/proj/build_output/_prebuild/mod_def_2.{mods_combo_hash}.mod')),
+            ],
+            any_order=True,
+        )
+
+    def ensure_mods_restored(self, mock_copy, mods_combo_hash='eac3b22d'):
+        # make sure previously built mod files were copied FROM the prebuilds folder
+        mock_copy.assert_has_calls(
+            calls=[
+                call(Path(f'/fab/proj/build_output/_prebuild/mod_def_1.{mods_combo_hash}.mod'),
+                     Path('/fab/proj/build_output/mod_def_1.mod')),
+                call(Path(f'/fab/proj/build_output/_prebuild/mod_def_2.{mods_combo_hash}.mod'),
+                     Path('/fab/proj/build_output/mod_def_2.mod')),
+            ],
+            any_order=True,
+        )
+
+    def test_without_prebuild(self):
+        # call compile_file() and return a CompiledFile
+        compiler, flags, analysed_file, expect_object_fpath = self.content()
+
+        with mock.patch('pathlib.Path.exists', return_value=False):  # no output files exist
+            with mock.patch('fab.steps.compile_fortran.CompileFortran.compile_file') as mock_compile_file:
+                with mock.patch('shutil.copy2') as mock_copy:
+                    res = compiler.process_file(analysed_file)
+
+        assert res == CompiledFile(input_fpath=analysed_file.fpath, output_fpath=expect_object_fpath)
+        mock_compile_file.assert_called_once_with(analysed_file, flags, output_fpath=expect_object_fpath)
+        self.ensure_mods_stored(mock_copy)
+
+    def test_with_prebuild(self):
+        # If the mods and obj are prebuilt, don't compile.
+        compiler, flags, analysed_file, expect_object_fpath = self.content()
+
+        with mock.patch('pathlib.Path.exists', return_value=True):  # mod def files and obj file all exist
+            with mock.patch('fab.steps.compile_fortran.CompileFortran.compile_file') as mock_compile_file:
+                with mock.patch('shutil.copy2') as mock_copy:
+                    res = compiler.process_file(analysed_file)
+
+        assert res == CompiledFile(input_fpath=analysed_file.fpath, output_fpath=expect_object_fpath)
+        mock_compile_file.assert_not_called()
+        self.ensure_mods_restored(mock_copy)
+
+    def test_file_hash(self):
```

Comment:
Some of these tests might be better if they used real files instead of mocking the existence of the expected filenames.

Reply:
We've made [a ticket](https://github.com/metomi/fab/projects/8#card-85865236) to think about this.